pkg/util/manager/manager.go

```diff
@@ -0,0 +1,106 @@
+package manager
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
+	workv1alpha1 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha1"
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+	"github.com/karmada-io/karmada/pkg/util"
+)
+
+// typeIndexerFuncMap contains internal type and its indexerFunc used to set up indexer.
+var typeIndexerFuncMap = map[client.Object]client.IndexerFunc{
+	&workv1alpha2.ResourceBinding{}:        indexForResourceBinding(),
+	&workv1alpha1.ResourceBinding{}:        indexForResourceBinding(),
+	&workv1alpha2.ClusterResourceBinding{}: indexForClusterResourceBinding(),
+	&workv1alpha1.ClusterResourceBinding{}: indexForClusterResourceBinding(),
+	&workv1alpha1.Work{}:                   indexForWork(),
+}
+
+type labelIndexerManager struct {
+	manager.Manager
+}
+
+// NewLabelIndexerManager returns a labelIndexerManager which set up field indexer for all needed internal type using typeIndexerFuncMap.
+// This file need to be refactored when controller-runtime support labelIndexer and client-go support multiIndexer in the future.
+// refer to:
+//		https://github.com/karmada-io/karmada/issues/2234
+//		https://github.com/kubernetes/kubernetes/pull/109334
+//      https://github.com/kubernetes-sigs/controller-runtime/pull/1838
+func NewLabelIndexerManager(config *rest.Config, options manager.Options) (manager.Manager, error) {
+	manager, err := manager.New(config, options)
+	if err != nil {
+		return nil, err
+	}
+
+	lim := &labelIndexerManager{
+		Manager: manager,
+	}
+
+	// register label indexer
+	ctx := context.Background()
+	for obj, indexerFunc := range typeIndexerFuncMap {
+		err := lim.Manager.GetFieldIndexer().IndexField(ctx, obj, "metadata.labels", indexerFunc)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return lim, nil
+}
+
+func indexForResourceBinding() client.IndexerFunc {
+	return func(object client.Object) []string {
+		return indexKeys(object, []string{
+			policyv1alpha1.PropagationPolicyNamespaceLabel,
+			policyv1alpha1.PropagationPolicyNameLabel,
+			policyv1alpha1.ClusterPropagationPolicyLabel,
+		})
+	}
+}
+
+func indexForClusterResourceBinding() client.IndexerFunc {
+	return func(object client.Object) []string {
+		return indexKeys(object, []string{
+			policyv1alpha1.ClusterPropagationPolicyLabel,
+		})
+	}
+}
+
+func indexForWork() client.IndexerFunc {
```

Comment:
When List work objects with label FederatedResourceQuotaNamespaceLabel and FederatedResourceQuotaNameLabel
https://github.com/karmada-io/karmada/blob/ad94a94dce24b98e59cdd9d5018a983ef0fa64ac/pkg/controllers/federatedresourcequota/federated_resource_quota_status_controller.go#L65-L72

According to your index function,if the work object has three labels it may be cause failure.
Eg:
workobject:
[
labels:
    "FederatedResourceQuotaNamespaceLabel":1,
    "FederatedResourceQuotaNameLabel":2,
    "ServiceNameLabel":3,
]
the index value is "FederatedResourceQuotaNamespaceLabel=1&FederatedResourceQuotaNameLabel=2&ServiceNameLabel=3",
but List function calculate the index value is "FederatedResourceQuotaNamespaceLabel=1&FederatedResourceQuotaNameLabel=2".
I'am afraid this will cause can't find the objects L in such sitaution.

Reply:
Yes. Is this possible that work both have the three labels above?